core/src/main/java/org/apache/iceberg/ManifestWriter.java

```diff
@@ -36,14 +38,33 @@
 
   static ManifestFile copyAppendManifest(ManifestReader reader, OutputFile outputFile, long snapshotId,
                                          SnapshotSummary.Builder summaryBuilder) {
+    return copyManifest(reader, outputFile, snapshotId, summaryBuilder, Sets.newHashSet(ManifestEntry.Status.ADDED));
+  }
+
+  static ManifestFile copyManifest(ManifestReader reader, OutputFile outputFile, long snapshotId,
+                                   SnapshotSummary.Builder summaryBuilder,
+                                   Set<ManifestEntry.Status> allowedEntryStatuses) {
     ManifestWriter writer = new ManifestWriter(reader.spec(), outputFile, snapshotId);
     boolean threw = true;
     try {
       for (ManifestEntry entry : reader.entries()) {
-        Preconditions.checkArgument(entry.status() == ManifestEntry.Status.ADDED,
-            "Cannot append manifest: contains existing files");
-        summaryBuilder.addedFile(reader.spec(), entry.file());
-        writer.add(entry);
+        Preconditions.checkArgument(
+            allowedEntryStatuses.contains(entry.status()),
+            "Invalid manifest entry status: %s (allowed statuses: %s)",
+            entry.status(), allowedEntryStatuses);
+        switch (entry.status()) {
+          case ADDED:
+            summaryBuilder.addedFile(reader.spec(), entry.file());
+            writer.add(entry);
+            break;
+          case EXISTING:
+            writer.existing(entry);
+            break;
+          case DELETED:
+            summaryBuilder.deletedFile(reader.spec(), entry.file());
```

Comment:
If the entry's snapshot ID is not the manifest's snapshot ID, the delete should be dropped.

Reply:
This one is related to other comments: I think it makes sense for utilities to read only live entries from manifests and make sure they change the status of entries to `EXISTING`. This method can stay general-purpose in case we need it later (we are already using it for appends and rewrites).

Reply:
`ManifestReader.entries` returns all entries unmodified.